src/main/java/uk/gov/ch/developer/docs/controller/developer/SignInController.java

```diff
@@ -0,0 +1,231 @@
+package uk.gov.ch.developer.docs.controller.developer;
+
+import java.io.IOException;
+import java.security.SecureRandom;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.apache.commons.codec.binary.Base64;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import com.nimbusds.jose.EncryptionMethod;
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWEAlgorithm;
+import com.nimbusds.jose.JWEHeader;
+import com.nimbusds.jose.JWEObject;
+import com.nimbusds.jose.Payload;
+import com.nimbusds.jose.crypto.DirectEncrypter;
+import net.minidev.json.JSONObject;
+import uk.gov.companieshouse.environment.EnvironmentReader;
+import uk.gov.companieshouse.environment.impl.EnvironmentReaderImpl;
+import uk.gov.companieshouse.logging.Logger;
+import uk.gov.companieshouse.logging.LoggerFactory;
+import uk.gov.companieshouse.session.Session;
+import uk.gov.companieshouse.session.SessionImpl;
+import uk.gov.companieshouse.session.SessionKeys;
+import uk.gov.companieshouse.session.handler.SessionHandler;
+import uk.gov.companieshouse.session.model.SignInInfo;
+import uk.gov.companieshouse.session.model.UserProfile;
+
+@Controller
+@RequestMapping("/signin")
+public class SignInController {
+
+    private static final String OAUTH_COMPANY_SCOPE_PREFIX =
+            "https://api.companieshouse.gov.uk/company/";
+
+    private static final EnvironmentReader reader = new EnvironmentReaderImpl();
```

Comment:
This probably should be an injected implementation.

Reply:
As part of the application.properties file?

Reply:
I was referring to the EnvironmentReader instance. I would expect the implementation to be a bean, rather than explicitly created. This gives us better options for testing and differing implementations if necessary for different environments.

Reply:
@Autowired basically.

Reply:
Similarly, where a `new SessionImpl` is performed lower in the code, it ought to be obtained from the factory, so that we can again mock it and replace it easily, although that won't be auto-wired, I suspect.

Reply:
SessionFactory doesn't have an implementation to return a brand new session directly.

Reply:
Implemented the java bean for the env reader